### PR TITLE
Manage the command line options of Apia with the ".apia" files.

### PR DESCRIPTION
**Pull request by jonaprieto**
This new feature helps to avoid writing in your command line every single value option that you often use or you probably gonna use the next n times.

For instance, if you use `metis` as your favorite ATP, consider add it to the .apia file:

``` yaml
# ~/.apia
atp: ["metis"]
```

Now, try it! all the possible configuration options are available. Check `apia --help`.

### CODE REVIEW

**Comment by asr** [comment]
I got the following error:

```
$ make install-bin
$ make agda_apia_changed
...
shelltest --color --precise \
                  test/succeed/non-conjectures/non-conjectures.test
:test/succeed/non-conjectures/non-conjectures.test:1: [Failed]
...
:test/succeed/non-conjectures/non-conjectures.test:5: [Failed]
...
         Test Cases  Total      
 Passed  7           7          
 Failed  2           2          
 Total   9           9          
make[1]: *** [non_conjectures] Error 1
```

**Comment by asr** [comment]
The files in `src/data` shouldn't be hidden files.

**Comment by asr** [comment]
Please add lower and upper bounds for the library dependencies in `apia.cabal`.

**Comment by asr** [comment]
How can a user knows about the new functionality implemented in this PR?

File: apia.cabal [hunk 1]
```diff
@@ -51,6 +51,8 @@ executable apia
                  -- tranformers 0.4.0.0 was deprecated.
                  , transformers >= 0.3 && < 0.4 || >= 0.4.1.0 && < 0.6
                  , unordered-containers >= 0.2.5.0 && < 0.3
+                 , yaml >= 0.8.17
+                 , deepseq >= 1.4.0
```
Comment (asr) [inline apia.cabal]: Keep the alphabetic order in the dependencies.

File: apia.cabal [hunk 2]
```diff
@@ -94,10 +96,11 @@ executable apia
...
 
-  ghc-options: -w
-               -Werror
-               -- Using cpphs as the C preprocessor.
-               -pgmP cpphs -optP --cpp
+  ghc-options:
+    -w
+    -Werror
+    -- Using cpphs as the C preprocessor.
+    -pgmP cpphs -optP --cpp
```
Comment (asr) [inline apia.cabal:108]: Why is the above change related to the `.apia` file?

Comment (jonaprieto) [inline apia.cabal:108]: I didn't touch it intentionally. Maybe was my editor.

File: apia.cabal [hunk 3]
```diff
@@ -163,7 +166,7 @@ executable apia
 
   if impl(ghc >= 8.0)
     ghc-options: -Wmissing-local-signatures
-                 -Wmissing-monadfail-instances
+                 -Wmissing-monadfail-instance
```
Comment (asr) [inline apia.cabal]: Why did you make this change?

File: apia.cabal [hunk 4]
```diff
@@ -179,4 +182,4 @@ executable apia
   -- enabled by default.
   ghc-options: -rtsopts
 
-  ghc-prof-options: -fprof-auto
+  ghc-prof-options: -fprof-auto
```
Comment (asr) [inline apia.cabal]: Missing newline.

File: src/Apia/.apia [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+snapshot-no-error: false
+snapshot-test: false
+spass-path: "SPASS"
+time: 2
+tptp2X-path:  "tptp2X"
+tptp4X-path: "tptp4X"
+unproven-conjecture-no-error: false
+vampire-path: "vampire_lin64"
+verbose: 20
+z3-path: "z3"
```
Comment (asr) [inline src/Apia/.apia]: Missing newline.

File: src/Apia/Defaults.hs [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+-- | Get the default values for the command-line 'Options'.
+getDefaults :: IO Options
+getDefaults = do
+  paths <- sequence [getCurrentDirectory, getHomeDirectory]
+  -- , (return.takeDirectory) __FILE__ ]
+  files <- filterM doesFileExist (map (</>".apia") paths)
+  -- print files
+  loaded <- mapM loadYAML files
+  return $ finalConfig loaded
```
Comment (asr) [inline src/Apia/Defaults.hs]: Missing newline.

File: src/Apia/Monad/Environment.hs [hunk 7]
```diff
@@ -20,15 +20,16 @@ module Apia.Monad.Environment ( env ) where
 import Apia.Prelude
 
 import Apia.Options     ( Options, processOptions )
+import Apia.Defaults     ( getDefaults )
```
Comment (asr) [inline src/Apia/Monad/Environment.hs]: Keep the importation in alphabetic order.

File: src/Apia/Monad/Environment.hs [hunk 8]
```diff
@@ -20,15 +20,16 @@ module Apia.Monad.Environment ( env ) where
...
 -- | The environment.
 env ∷ IO Options
 env = do
   args ← getArgs
-  case processOptions args of
+  defaults ← getDefaults
+  case processOptions args defaults of
     Left err → die err
-    Right o  → return o
+    Right o  → return o
```
Comment (asr) [inline src/Apia/Monad/Environment.hs]: Missing newline.

File: src/Apia/Options.hs [hunk 9]
```diff
@@ -46,10 +45,12 @@ module Apia.Options
            , optWithIleanCoP
            , optWithMetis
            , optWithSPASS
+           , optWithtptp2X
```
Comment (asr) [inline src/Apia/Options.hs]: Please create a different PR for adding this missing option.

File: src/Apia/Options.hs [hunk 10]
```diff
@@ -408,7 +383,7 @@ options =
   ]
 
 usageHeader ∷ String → String
-usageHeader prgName = "Usage: " ++ prgName ++ " [OPTIONS] FILE\n"
+usageHeader prgName = "\nApia\nUsage: " ++ prgName ++ " [OPTIONS] FILE\n"
```
Comment (asr) [inline src/Apia/Options.hs]: The program name shouldn't be hard-coded in the source files.

File: src/Apia/Options.hs [hunk 11]
```diff
@@ -423,5 +398,5 @@ processOptionsHelper argv f defaults =
     (_, _, errs) → Left $ pretty $ initDef (__IMPOSSIBLE__) $ init $ unlines errs
 
 -- | Processing the command-line 'Options'.
-processOptions ∷ [String] → Either Doc Options
-processOptions argv = processOptionsHelper argv inputFileOpt defaultOptions
+processOptions ∷ [String] → Options -> Either Doc Options
+processOptions argv = processOptionsHelper argv inputFileOpt
```
Comment (asr) [inline src/Apia/Options.hs]: Missing newline.

File: apia.cabal (hunk 2 above)
Comment (asr) [inline apia.cabal:108]: So, revert the changes please.

File: src/Apia/Options.hs (hunk 9 above)
Comment (asr) [inline src/Apia/Options.hs]: ~~Please create a different PR for adding this missing option.~~
`tptp2X` is not currently used. Are you working with an update version of the repository?

File: src/Apia/Monad/Environment.hs (hunk 8 above)
Comment (asr) [inline src/Apia/Monad/Environment.hs]: You can run `make check-whitespace` for avoiding these problems.

File: src/Apia/Options.hs (hunk 9 above)
Comment (jonaprieto) [inline src/Apia/Options.hs]: This is  weird, I can actually see the commit that you did about that, I pulled not so long ago, and this were the last commit and after them, I have my commits.

``` bash
commit fe58c1c9aeaf9c02d598918421197698fbdc9ffd
Author: Andrés Sicard-Ramírez <asr@eafit.edu.co>
Date:   Wed Jul 20 16:11:44 2016 -0500

    Supported TPTP 6.4.0.

commit cbf5cad8bb7ee92c8eaf46de7dff583b512cd437
Author: Andrés Sicard-Ramírez <asr@eafit.edu.co>
Date:   Wed Jul 20 15:12:05 2016 -0500

    Using `tptp4X` instead of `tptp2X`.
```

What should I do?

**Comment by jonaprieto** [comment]
Just pending a new section in the README about .apia files.

**Comment by asr** [comment]
Note that your PR didn't pass the CI test.

File: src/Apia/Options.hs (hunk 9 above)
Comment (asr) [inline src/Apia/Options.hs]: Do you see the change if you squash all your commits in only one commit?

**Comment by asr** [comment]
I run `fix-whitespace` and `hlint` before creating a commit. See the instructions in the new [HACKING.md](https://github.com/asr/apia/blob/master/HACKING.md) file.

**Comment by asr** [comment]
Running `make hlint` I got the following problems:

```
src/Apia/Defaults.hs:344:16: Suggestion: Redundant $
Found:
  doesFileExist $ dotapia
Why not:
  doesFileExist dotapia

src/Apia/Defaults.hs:345:3: Warning: Use when
Found:
  if not ishomeapia then
    do dotapiatemplate <- getDataFileName "apia_template.yml"
       copyFile dotapiatemplate dotapia
    else return ()
Why not:
  Control.Monad.when (not ishomeapia) $
    do dotapiatemplate <- getDataFileName "apia_template.yml"
       copyFile dotapiatemplate dotapia

2 hints
make: *** [hlint] Error 1
```

**Comment by asr** [comment]
I just added the HLint test to Travis.

**Comment by jonaprieto** [comment]
I got this:

``` bash
hlint --color=never \
              --cpp-file=dist/build/autogen/cabal_macros.h \
              --cpp-include=src/Apia/ \
              src/
src/Apia/Translation.hs:85:8: Error: Parse error: role
Found:
    toAFor ∷ TPTPRole → QName → Definition → T AF
  > toAFor role qName def = do
      let ty ∷ Type
          ty = defType def

src/Apia/TPTP/Files.hs:124:24: Error: Parse error: role
Found:
    agdaOriginalTerm ∷ QName → TPTPRole → Text
  > agdaOriginalTerm qName role =
      "% The Agda term was:\n"
      +++ "% Name: " +++ (T.pack . prettyShow . qnameToConcrete) qName +++ "\n"

2 hints
Makefile:469: recipe for target 'hlint' failed
```

I don't know what is wrong.

**Comment by asr** [comment]
> I don't know what is wrong.

You are using an old version of HLint (see the `Makefile` for the required version).

**Comment by asr** [comment]
I patch verifying the minimal version of HLint is welcome :-)

**Comment by asr** [comment]
> I run `fix-whitespace` and `hlint` before creating a commit

I also added the `fix-whitespace` test to Travis.

File: README.md [hunk 12]
```diff
@@ -114,6 +114,52 @@ Command-line options
...
+when running Apia.
+
+### Examples
+
+We want to use the ATPs E and Metis with Apia. Also, we want to save the
+results in the directory `/myoutputdir` and use the option `check`.
+Then, our Apia file should be similar to this one.
+
+```yaml
+# file ~/.apia
```
Comment (asr) [inline README.md]: I guess you meant `cat` instead of `file`. Please fix it.

File: README.md [hunk 13]
```diff
@@ -114,6 +114,52 @@ Command-line options
...
+atp: ["e", "metis"]
+output-dir: "/myoutputdir"
+check: true
+```
+
+*Note:* Apia always saves an `.apia` file in the home directory with the basic
+settings if it doesn't exist. If for some reason, this file `~/.apia` disappears,
+the next time you run Apia, it will create an `.apia` file for you in your home
```
Comment (asr) [inline README.md]: Is the  `.apia` file created after the installation or after using Apia for first time?

File: apia.cabal [hunk 14]
```diff
@@ -34,23 +39,25 @@ executable apia
...
+           -- package versioning policy.
+           , pretty >= 1.1.1.0 && < 1.1.1.2 || >= 1.1.2 && < 1.2
+           , process >= 1.1.0.2 && < 1.5
+           , regex-compat >= 0.95.1 && < 0.96
+           , safe >= 0.3.5 && < 0.4
+           , text >= 0.11.3.1 && < 1.3
+           -- tranformers 0.4.0.0 was deprecated.
+           , transformers >= 0.3 && < 0.4 || >= 0.4.1.0 && < 0.6
+           , unordered-containers >= 0.2.5.0 && < 0.3
+           , yaml >= 0.8.17 &&  < 0.9
```
Comment (asr) [inline apia.cabal:60]: Please fix the indentation of the dependencies.

File: apia.cabal [hunk 15]
```diff
@@ -59,6 +66,7 @@ executable apia
     Apia.CheckTPTP
     Apia.Common
     Apia.Dump
+    Apia.Defaults
```
Comment (asr) [inline apia.cabal:67]: Keep the modules in alphabetic order.

File: src/Apia/ATPs.hs [hunk 16]
```diff
@@ -306,12 +302,12 @@ smt2Ext = ".smt2"
...
 selectedATPs = do
-  atpsAux ← askTOpt optATP
-
-  let atps ∷ [String]
-      atps = if null atpsAux then defaultATPs else atpsAux
-
-  mapM optATP2ATP atps >>= modifyTATPs
+  atps ← askTOpt optATP
+  let msgError ∷ Doc
+      msgError = pretty "At least you need to specified one valid ATP"
```
Comment (asr) [inline src/Apia/ATPs.hs]: Please add a test case for this error.

File: src/Apia/Common.hs [hunk 17]
```diff
@@ -40,4 +40,4 @@ instance Pretty ATP where
 -- | Target languages.
 data Lang = TPTP   -- ^ FOF (First-order form).
           | SMT2   -- ^ SMT-LIB v2.
-          deriving Eq
+          deriving (Eq, Show)
```
Comment (asr) [inline src/Apia/Common.hs:43]: Why this change was required?

File: src/Apia/Defaults.hs [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+
+{-# LANGUAGE CPP               #-}
+{-# LANGUAGE MultiWayIf        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE UnicodeSyntax     #-}
+
+module Apia.Defaults (
+    getDefaults,
+    )
+where
```
Comment (asr) [inline src/Apia/Defaults.hs]: Better `module Apia.Defaults ( getDefaults ) where`.

File: src/Apia/Defaults.hs [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+
+type Key = T.Text
+
+data Config = Config Object
+  deriving (Eq, Show)
+
+instance NFData Config where
+  rnf (Config o) = rnf o `seq` ()
+
+look ∷ FromJSON b => Key → Config → Maybe b
```
Comment (asr) [inline src/Apia/Defaults.hs]: Use the Unicode character `⇒`.

File: src/Apia/Defaults.hs [hunk 20]
```diff
@@ -0,0 +1,359 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      : Apia.Defaults
+-- Copyright   : (c) Andrés Sicard-Ramírez 2009-2015
+-- License     : See the file LICENSE.
+--
+-- Maintainer  : Jonathan Prieto-Cubides <jprieto9@eafit.edu.co>
+-- Stability   : experimental
+--
+-- Get the default values argumentsa.
```
Comment (asr) [inline src/Apia/Defaults.hs]: Fix the typo.

File: src/Apia/Defaults.hs [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
+-----------------------------------------------------------------------------
+
+{-# LANGUAGE CPP               #-}
+{-# LANGUAGE MultiWayIf        #-}
```
Comment (asr) [inline src/Apia/Defaults.hs:14]: Nice!

File: src/Apia/Defaults.hs [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+      _           → opts
+
+setLang ∷ Options → Config → Options
+setLang opts config = setNoInternalEquality newopts config
+  where
+    newopts ∷ Options
+    newopts = case look "lang" config of
+      (Just  val)  → if | val == T.pack "tptp" → opts { optLang = TPTP }
+                         | val == T.pack "smt"  → opts { optLang = SMT2 }
+                         | otherwise            → opts
```
Comment (asr) [inline src/Apia/Defaults.hs]: Fix indentation.

File: src/Apia/Defaults.hs [hunk 23]
```diff
@@ -0,0 +1,359 @@
...
+      _           → opts
```
Comment (asr) [inline src/Apia/Defaults.hs]: Fix indentation.

File: src/Apia/Defaults.hs (hunk 19 above)
Comment (asr) [inline src/Apia/Defaults.hs]: Rename the function. (`look` is too common).

File: src/Apia/Defaults.hs [hunk 24]
```diff
@@ -0,0 +1,359 @@
...
+
+setVals ∷ Options → Config → Options
+setVals = setATP
+
+setATP ∷ Options → Config → Options
+setATP opts config = setCheck newopts config
+  where
+    newopts ∷ Options
+    newopts = case look "atp" config of
+      (Just  val) → opts { optATP = val }
```
Comment (asr) [inline src/Apia/Defaults.hs]: Better `Just val`.  HLint should report it. Please create an issue on HLint CC'ing me.

File: src/Apia/Defaults.hs [hunk 25]
```diff
@@ -0,0 +1,359 @@
...
+  where
+    hs ∷ [Object]
+    hs = map getHashMap cfs
+    getHashMap ∷ Config → Object
+    getHashMap (Config o) = o
+
+loadYAML ∷ FilePath → IO Config
+loadYAML dotapia = do
+    decoded ←  YamlInclude.decodeFile dotapia
+    return $ Config $ fromMaybe HashMap.empty decoded
```
Comment (asr) [inline src/Apia/Defaults.hs]: Use only 2 spaces for the indentation, please.

File: src/Apia/Defaults.hs [hunk 26]
```diff
@@ -0,0 +1,359 @@
...
+-- | it creates a new one based on .apia_tempalte file.
+checkApiaHomeFile ∷ IO ()
+checkApiaHomeFile = do
+  homedir ← getHomeDirectory
+  let dotapia ∷ FilePath
+      dotapia = homedir </> ".apia"
+  ishomeapia ← doesFileExist dotapia
+  unless ishomeapia $ do
+    dotapiatemplate ← getDataFileName "apia_template.yml"
+    copyFile dotapiatemplate dotapia
```
Comment (asr) [inline src/Apia/Defaults.hs]: Use Camel-case please.

File: src/Apia/Defaults.hs [hunk 27]
```diff
@@ -0,0 +1,356 @@
...
+checkApiaHomeFile ∷ IO ()
+checkApiaHomeFile = do
+  homeDir ← getHomeDirectory
+  let dotApia ∷ FilePath
+      dotApia = homeDir </> ".apia"
+  isHomeApia ← doesFileExist dotApia
+  unless isHomeApia $ do
+    dotApiaTemplate ← getDataFileName "apia_template.yml"
+    copyFile dotApiaTemplate dotApia
+
```
Comment (asr) [inline src/Apia/Defaults.hs:375]: Do not hard-coded the files names in the source. Use a variable instead.

File: src/Apia/Defaults.hs [hunk 28]
```diff
@@ -0,0 +1,356 @@
...
+    getHashMap ∷ Config → Object
+    getHashMap (Config o) = o
+
+loadYAML ∷ FilePath → IO Config
+loadYAML dotApia = do
+  decoded ←  YamlInclude.decodeFile dotApia
+  return $ Config $ fromMaybe HashMap.empty decoded
+
+-- | If the .apia file doesn't exist in the home directory,
+-- | it creates a new one based on apia_tempalte.yml file.
```
Comment (asr) [inline src/Apia/Defaults.hs]: Fix the typo.

File: README.md [hunk 29]
```diff
@@ -114,6 +114,52 @@ Command-line options
...
+*Note:* When Apia is invoked outside a project it will source project specific
+options from `~/.apia`.  Options in this file will be ignored for a project with
+its own `<project dir>/.apia`.
+
+### Project-specific config
+
+Project-specific options are only valid in the `.apia` file local to a
+project, not in the user config files.
+
+> Note: We define **project** to mean a directory that contains an `.apia`
```
Comment (asr) [inline README.md]: Better '_project_' than '**project**'.

File: README.md [hunk 30]
```diff
@@ -114,6 +114,52 @@ Command-line options
...
+> file, which specifies valid options. The options are specified in the `help`
+> command. Check `apia --help` to see all options available.
+
+In your project-specific options, you specify  **which options to use**
```
Comment (asr) [inline README.md]: Better '_which options to use_' than '**which options to use**'.

**Comment by asr** [comment]
Which is the function of the `src/data/apia.yml` file? That is, why is it necessary both .yml files?

File: src/Apia/Common.hs (hunk 17 above)
Comment (asr) [inline src/Apia/Common.hs:43]: What means the :+1: ? Are you going to answer the question?

File: src/Apia/Defaults.hs [hunk 31]
```diff
@@ -0,0 +1,356 @@
...
+
+setVals ∷ Options → Config → Options
+setVals = setATP
+
+setATP ∷ Options → Config → Options
+setATP opts config = setCheck newOpts config
+  where
+    newOpts ∷ Options
+    newOpts = case lookConfig "atp" config of
+      Just  val → opts { optATP = val }
```
Comment (asr) [inline src/Apia/Defaults.hs]: Remove the extra space.

**Comment by jonaprieto** [comment]
With apia.yml we set the global default options.
With apia_template.yml, we set the user default options.

**Comment by asr** [comment]
I see. Why not to use the same file + the difference between both files?

**Comment by asr** [comment]
Create a different PR for commits https://github.com/asr/apia/pull/34/commits/ec38aa1ec7e8b6c07ba9fc21c90873560694fb5d and https://github.com/asr/apia/pull/34/commits/b43eb7e5806a1182f65f46d03b90d42b9f5f6033 please.

File: src/Apia/Defaults.hs [hunk 32]
```diff
@@ -97,7 +97,7 @@ setCheck opts config = setDumpTypes newOpts config
   where
     newOpts ∷ Options
     newOpts = case lookConfig "check" config of
-      Just  val → opts { optCheck = val}
+      Just  val → opts { optCheck = val }
```
Comment (asr) [inline src/Apia/Defaults.hs]: The previous space was right. I was talking about the spaces between `Just` and `val`.

**Comment by jonaprieto** [comment]
The difference is basically the comments in the YAML, thinking in the final user. So, Do you want to change it? Use one file is ok for me.

**Comment by asr** [comment]
Please use one file and append the comments for the final user.

**Comment by asr** [comment]
Thinking again about the `.apia` file, this file _shouldn't_ be installed by default. If the user run `apia` without installing this file and without using the `--atp` option, `apia` should generate error.

```
"At least you need to specified one ATP"
```

File: src/Apia/ATPs.hs (hunk 16 above)
Comment (asr) [inline src/Apia/ATPs.hs]: Remove `valid`.

**Comment by asr** [comment]
FYI, I added the following tests to Travis:

```
  - make generated_all
  - make non_conjectures
  - make errors
  - make type_check_notes
  - make ATPs="e" prove_notes
```

**Comment by jonaprieto** [comment]
With last test you added to Travis I'll fail with .apia feature unless you add to Travis the ATP by default.

``` bash
λ ~/apia/ master* make generated_all
make generated_fol_theorems
make[1]: Entering directory '/home/jonaprieto/apia'
rm -r -f /tmp/apia
make generated_fol_theorems_aux
make[2]: Entering directory '/home/jonaprieto/apia'
Comparing test/succeed/fol-theorems/Agda/InternalTerms/DefTerm1.agda
apia: At least you need to specify one valid ATP
Makefile:95: recipe for target 'test/succeed/fol-theorems/Agda/InternalTerms/DefTerm1.generated_fol_theorems' failed
make[2]: *** [test/succeed/fol-theorems/Agda/InternalTerms/DefTerm1.generated_fol_theorems] Error 1
make[2]: Leaving directory '/home/jonaprieto/apia
```

Then, I suggest that after install Apia, in .travis.yml add something like `echo "atp: [\"e\"]" > ~/.apia`.
I got this error because you had E as a default ATP, and now with my last modification. There is no default ATP.

**Comment by asr** [comment]
Thanks for not install the `.apia` file by default. Remember to update the README.

**Comment by asr** [comment]
You can fix `make generated_all` on Travis using something like

```
make APIA="apia --atp=e --check" generated_all
```

File: src/Apia/Defaults.hs [hunk 33]
```diff
@@ -0,0 +1,365 @@
...
+
+apiaNameFile ∷ FilePath
+apiaNameFile = ".apia"
+
+apiaTemplate :: FilePath
+apiaTemplate = "apia.yml"
+
+-- | If the .apia file doesn't exist in the home directory,
+-- | it creates a new one based on apiaTemplate file.
+checkApiaHomeFile ∷ IO ()
```
Comment (asr) [inline src/Apia/Defaults.hs]: This function shouldn't be necessary.

File: src/Apia/Defaults.hs [hunk 34]
```diff
@@ -0,0 +1,365 @@
...
+      _        → opts
+
+setWithZ3 ∷ Options → Config → Options
+setWithZ3 opts config = newOpts
+  where
+    newOpts ∷ Options
+    newOpts = case lookConfig "with-z3" config of
+      Just val → opts { optWithZ3 = val }
+      _        → opts
+
```
Comment (asr) [inline src/Apia/Defaults.hs:314]: Is it strictly necessary that each `setFoo` function (except the "first" one) depends of the previous one?

Comment (jonaprieto) [inline src/Apia/Defaults.hs:314]: No really. Do you want to change it?

Comment (asr) [inline src/Apia/Defaults.hs:314]: Yes, please.

**Comment by asr** [comment]
In a fresh clone of your PR, I got the error you pointed out [above](https://github.com/asr/apia/pull/34#issuecomment-239658986). You can fix the error by replacing in the `Makefile`

```
APIA = dist/build/apia/apia --check
```

by

```
APIA = dist/build/apia/apia --check --atp=e
```

The above change also fix the current Travis error on `make generated_all`.

**Comment by asr** [comment]
Please note the Travis doesn't send (email) notifications on pull requests.

**Comment by jonaprieto** [comment]
Please, look at this. The output errors after run make agda_apia_changed
https://gist.github.com/jonaprieto/6899cbcdc2ca928aa96fc10b039301cb
the Makefile is https://gist.github.com/jonaprieto/671a566879f7d48f3f006a11e9fcafa2

**Comment by asr** [comment]
> Please, look at this. The output errors after run make agda_apia_changed

I fixed (in master) the problem by 5bee895.

File: src/Apia/Defaults.hs [hunk 35]
```diff
@@ -0,0 +1,385 @@
...
+  , setWithCVC4
+  , setWithE
+  , setWithEquinox
+  , setWithIleanCoP
+  , setWithMetis
+  , setWithSPASS
+  , setWithtptp4X
+  , setWithVampire
+  , setWithZ3
+  ]
```
Comment (asr) [inline src/Apia/Defaults.hs:345]: :+1:

**Comment by asr** [comment]
Thanks for implementing this feature.

File: src/Apia/Common.hs (hunk 17 above)
Comment (asr) [reply src/Apia/Common.hs:43]: See 483dd61f15cd97ef33ec49c9785a1370f751af66.
